### PR TITLE
Massive RCD Port & Updates

### PR DESCRIPTION
**Pull request by r4d6**
## About The Pull Request
This port a _lot_ of stuff from TG.

- https://github.com/tgstation/tgstation/pull/47894
- https://github.com/tgstation/tgstation/pull/48362
- https://github.com/tgstation/tgstation/pull/43598
- https://github.com/tgstation/tgstation/pull/46801
- https://github.com/tgstation/tgstation/pull/45357
- https://github.com/tgstation/tgstation/pull/47289 (More like just make RCD and RPD printable from Engineering's protolathe)

## Why It's Good For The Game
*Clear throat*
1. Animations for the RCD is cool as fuck, and fix said animation being pulled by Lord Singuloth
2. Make Reinforced doors impossible to deconstruct, giving another reason to do so.
3. You were able to remove floors without costing any Matter-Units, no longer is this the case
4. Allow you to make APCs, Firelocks and Fire alarms with RCDs. Also change the upgrades to be a flag and add another disk to make them.
5. Allow Engineers to print fully loaded RCDs from their protolathe for an increased cost and RPDs for the same cost as in an hacked autolathe, without removing them from said lathe.

## Changelog
:cl:
add: Animations for the RCD
fix: Said animations being pulled by the Singulo
balance: Prevent Reinforced Doors from being RCD'ed
fix: Deconstructing Floors now cost MUs like everything else
tweak: changed the RCD's upgrades into flags
add: Allow RCDs to print APC, Firelocks and Fire Alarms with the right upgrade
balance: Allow Engineers to print RCDs and RPDs from their protolathe without needing to hack one.
/:cl:

### CODE REVIEW

**Comment by GrayRachnid** [comment]
>Make Reinforced doors impossible to deconstruct, giving another reason to do so.

Engiborgs and Mech RCD can deconstruct R-walls, would they be stopped by reinforced doors too?

**Comment by r4d6** [comment]
At that point does it even matter?

Also the Mech RCD doesn't give a shit about what it remove anyway.

**Comment by GrayRachnid** [comment]
>At that point does it even matter?

It matters if the door is depowered(by the apc) and bolted.
RCDing a door is cheaper than an r-wall. I don't think Engiborg RCDs should be hard blocked from RCDing r-doors.

>Also the Mech RCD doesn't give a shit about what it remove anyway.

Do you mean it has no restrictions at all?

**Comment by r4d6** [comment]
1. The mech don't have any check if it is a r-wall or not, it can even RCD at distance!
2. Even if the Engie Borgs can't RCD R-doors, they have AA, have the tools to remove the protection, then RCD. They can also just RCD the wall right next to the door, even if it is a R-Wall.

**Comment by r4d6** [comment]
Also, do you mind if we take this discussion on Discord?

File: code/modules/research/designs/misc_designs.dm [hunk 1]
```diff
@@ -419,6 +429,26 @@
...
+	category = list("Equipment")
+	departmental_flags =  DEPARTMENTAL_FLAG_ENGINEERING
+
+/datum/design/rpd
+	name = "Rapid Pipe Dispenser"
+	desc = "A tool that can construct and deconstruct pipes on the fly."
+	id = "rpd"
+	build_type = PROTOLATHE
+	materials = list(MAT_METAL = 75000, MAT_GLASS = 37500)
+	build_path = /obj/item/construction/rcd/loaded
```
Comment (Trilbyspaceclone) [inline code/modules/research/designs/misc_designs.dm]: Loaded?

Comment (r4d6) [reply code/modules/research/designs/misc_designs.dm]: Basically it cost more than in the Lathe, but come already filled with Matter-Units.

Comment (Trilbyspaceclone) [reply code/modules/research/designs/misc_designs.dm]: How is the RPD loaded?

Comment (r4d6) [reply code/modules/research/designs/misc_designs.dm]: Sorry, going to fix that now

Comment (r4d6) [reply code/modules/research/designs/misc_designs.dm]: Done

**Review by Linzolle — APPROVED** [review]

**Comment by Linzolle** [comment]
code looks fine at least. Not gonna merge on just my approval though.

**Comment by r4d6** [comment]
I want it to be test-merged first anyway

**Comment by Detective-Google** [comment]
> > At that point does it even matter?
> 
> It matters if the door is depowered(by the apc) and bolted.
> RCDing a door is cheaper than an r-wall. I don't think Engiborg RCDs should be hard blocked from RCDing r-doors.
> 
> > Also the Mech RCD doesn't give a shit about what it remove anyway.
> 
> Do you mean it has no restrictions at all?

you could just, y'know, cut out the reinforcements.

**Comment by r4d6** [comment]
You could remove the reinforcement, just like you can RCD the wall right next to it.

**Comment by BlackMajor** [comment]
Give the construction effect mouse_opacity 0 if you don't want it to be clickable or show up in right click menus.

**Comment by r4d6** [comment]
Done

**Pull request merged**